### PR TITLE
Generate text corpus as jsonl

### PR DESCRIPTION
**Pull request by quadrismegistus**
An alternative approach to text corpus generation: two files:
* metadata.json
* pages.jsonl.gz

### CODE REVIEW

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:41]: let Solr do this work for you - in parasolr queryset the method is `only` (inspired by django queryset) https://parasolr.readthedocs.io/en/latest/codedocs.html#parasolr.query.queryset.SolrQuerySet.only

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:36]: why do you need to set limits every iteration? should this be an update call? (could be I've forgotten how to use progressbar, it's been a while)

Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py:36]: while I'm thinking about it: would be good to make progressbar optional, since it's possibly we'll want to run this script without that (e.g. if we set it up as a cron job)

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:57]: I missed that you were generating the metadata export in this script as well. Would like to align with the admin export (minus internal fields, but I guess that can be a later step.

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:63]: why is the enumeration and doc limit filtering needed?

would likely be more efficient to iterate and save in chunks

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py:57]: I don't think the metadata is large enough to merit jsonl and don't think it needs to be the same

**Review by rlskoeser — COMMENTED** [review]
Thanks for the quick spike on this - I'd be interested in taking the output for a spin; I think jsonl might be a good option for us, more efficient to generate and work with.  

How do we make the call / who should make the call?

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:36]: the `set_limits` is from parasolor as method of the SolrQuerySet. progressbar ended up being simple and nearly identical to tqdm. will add an verbose/quiet flag

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:57]: Is the solr and admin metadata (or another export command) guaranteed to be in sync?

Agreed, metadata can be json easily, will change back.

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:63]: jsonlines by default saves by row, I'm not aware of chunking. The doclimit param is a little silly, legacy of previous script, I don't see why anyone would want a partial export where for the gensim script it made more sense, maybe we can just can it.

Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py]: If we need a doclimit, I think it should be applied at the solr query level - maybe a page limit would make more sense for the new version.  How long does the script take? Any value in a shorter run for dev/test? If not I agree about dropping the option.

File: ppa/archive/tests/test_generate_textcorpus.py
Comment (rlskoeser) [reply ppa/archive/tests/test_generate_textcorpus.py:59]: yes! we have a pytest plugin: https://github.com/Princeton-CDH/parasolr/blob/main/parasolr/pytest_plugin.py#L160

maybe it isn't included in the docs yet? (if you can confirm missing please create an issue or fix if it's easy; please note/suggest any other doc improvements, your perspective would be helpful)

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py]: ah, I forgot about that syntax because I never use it directly. SolrQuerySet is designed to support indexing like a list, so you should be able to do `[100:200]` to get a chunk of results and it should query Solr for you.  

quite possible this is not documented or not clearly documented (please flag/help resolve)

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:144]: cool

**Review by rlskoeser — COMMENTED** [review]
Started testing but discovered I didn't have page data in my dev solr collection.   I had some problems running it but I think they all are due to zero page results from Solr.  Successfully running it now with partial page data.

Preliminary comments:
- could we shorten the name of the manage script? 
- need to add orjsonl to requirements now that we've decided to go with that
- script crashes when there are no pages in solr; I know we are relying on the indexing but quick sanity check on page content would be helpful. Ideal would be page count in solr matches (or is close to) the total number of pages reported in the postgres database (now I'm curious how closely those two things match....). 

Looks like everything is working well, I'll reindex pages locally overnight and test a little tomorrow and let you know if I have any more feedback.

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:119]: I think `filter` might be slightly better than `search` based on how Solr caches things, although it might not make a difference here.

```suggestion
        qset = self.query_set.filter(item_type=item_type).order_by("id")
```

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:161]: I had been thinking of using the aliased solr queryset (which lets you rename fields)  in parasolr, which now has a method for updating the result document before it returns. I kind of suspect your generator solution is more efficient, though.

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:160]: I don't think this assertion is needed; if it does fail, the whole script will abort - but it also shouldn't fail! Can we make the assumptions clearer in the higher level comments or docstring?
```suggestion
```

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:204]: What controls whether it is gzipped or not? is it just based on the filepath?

If it's easy, would be nice to have a flag to generate plain jsonl - seems useful for dev at least.

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:141]: can we give this a more semantic name? would `chunks` or `chunk_iterator` make sense here?

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:37]: any reason you didn't use a dict? (that's what we use in the aliased queryset in parasolr)

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py:62]: I see the `sort_keys` option on the `json.dump` call now, new to me (and glad to know about it!). What order do we get without that?

**Review by rlskoeser — COMMENTED** [review]
@quadrismegistus this is fantastic work. I have a few small suggestions for the existing code, but the approach looks great. From a code perspective jsonl seems like a good solution.

From the debug logging it looks like parasolr is hitting Solr twice for every chunk of results; I suspect that is a parasolr problem because I don't see anything in your code that seems like it would cause that. 

I'm getting an error from Solr when I try to export all pages, have you seen anything like this?  (Error is parasolr request, not the new code)
```
requests.exceptions.JSONDecodeError: Invalid control character at: line 55045 column 5572 (char 20851707)
```

Please update this PR to point to develop when you get a chance.

LMK what you need from me on the unit tests so we can get this merged and delivered for acceptance testing.

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:108]: yep, that would work just as well

File: manage.py
Comment (quadrismegistus) [reply manage.py:14]: woops -- I didn't mean to commit any changes to manage.py

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:62]: Could be by dictionary key entry time/order if on a modern Python3 branch. It's not a big issue for our current work, just a practice I generally maintain. But you're right, consistency trumps that

Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py:62]: I appreciate the attention to the order of the output, I think that's valuable! Maybe there's a different way to manage that besides alpha key

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:125]: yes! I've been meaning to implement this. what is the field to query on there? I didn't see it in the solr original fields:

```
 "_version_": 1757837811292045312,
   "author": "Sheridan, Thomas, 1719-1788",
   "author_exact": "Sheridan, Thomas, 1719-1788",
   "cluster_id_s": "sheridancompleteSV",
   "collections": [
     "Dictionaries",
     "Linguistic"
   ],
   "collections_exact": [
     "Dictionaries",
     "Linguistic"
   ],
   "group_id_s": "CB0126082904",
   "id": "CB0126082904",
   "item_type": "work",
   "last_modified": "2023-02-14T19:59:33.011Z",
   "order": 0,
   "pub_date": 1796,
   "pub_place": "Philadelphia ",
   "publisher": "W. Young, Mills & Son, booksellers and stationers, no. 52, Second-Street, corner of Chesnut-Street",
   "sort_title": "complete dictionary of the English language, both with regard to sound and meaning: one main object of which is, to establish a plain and permanent standard of pronunciation. To which is prefixed, a prosodial grammar. By Thomas Sheridan, A.M.",
   "source_id": "CB0126082904",
   "source_t": [
     "Gale"
   ],
   "source_url": "https://link.gale.com/apps/doc/CB0126082904/ECCO?u=prin77918&sid=gale_api&xid=a5b78388",
   "subtitle": "both with regard to sound and meaning: one main object of which is, to establish a plain and permanent standard of pronunciation. To which is prefixed, a prosodial grammar. By Thomas Sheridan, A.M.",
   "title": "A complete dictionary of the English language,",
   "work_type_s": "full-work"
```

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:160]: Sure. fwiw this works ok, so at least we know it's a good assumption for the current solr indexing logic. we could add some notes about what this script expects to exist and in what form when we pull from solr, and just pull `d["source"][0]` if we want

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:204]: Yes, just based on filename— we could either add a `--pages-filename` argument (right now we only have an output directory flag) or a `--uncompressed` argument?

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:37]: oh ok, didn't realize— I thought parasolr wanted a list.

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py:125]: wow, the danger of trusting my memory. I just checked, you don't need to filter in solr because it isn't there! when a work is suppressed we index a stub record (id only) to make sure we replace any existing data. Sorry for misleading!

I think this does make the check on totals more important, though (# unsuppressed works, # pages expected in the db, actuals in the Solr index). I think I've done something like that elsewhere and can dig it up so we can adapt.

Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:125]: Just confirming: neither suppressed works nor the pages of suppressed works are in solr?

Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py:125]: Correct, that is what we expect (as long as indexing is in sync with the db). Want to make a note in the comments in the code as a reminder?

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py:204]: Makes more sense to me to make the uncompressed a flag vs filename, I think we want that to stay the same. Shorter is better, what about `--no-zip` or similar?

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py]: you're generating the correct syntax! parasolr uses the dict to generate that syntax and a few other things: https://github.com/Princeton-CDH/parasolr/blob/main/parasolr/query/aliased_queryset.py#L20-L28

aliased queryset isn't in the parasolr docs either, is it? 🤦‍♀️ 😞

**Comment by rlskoeser** [comment]
@quadrismegistus for reporting on / comparing total number of unsuppressed works and pages in solr and db, we have some similar code in the index pages manage command:

https://github.com/Princeton-CDH/ppa-django/blob/main/ppa/archive/management/commands/index_pages.py#L97-L98

`DigitizedWork.items_to_index` returns a django queryset of items that should be indexed, so you can call `.count()` on that for a total. `Page.total_to_index()` runs an aggregation query to sum all the page counts on the items to be indexed.

At the end of the index_pages script we report on the totals of items by type in solr:
https://github.com/Princeton-CDH/ppa-django/blob/main/ppa/archive/management/commands/index_pages.py#L122-L132

At the beginning of the corpus export script, it would be great to output the totals for both and warn loudly if they don't match. They _should_ match but we'll have to see whether or not they do actually match in practice - I just checked and my page counts are somewhat different in local dev, but that could be a db/pairtree mismatch.

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:10]: suggested revision

```suggestion
The full text corpus is generated from Solr; this script makes the following assumptions:
    * Suppressed works and pages are not indexed (expected behavior, if db and Solr are synchronized)
    * Each work has a single source (Gale or Hathi) even though it is indexed in Solr as a list
    
```

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:138]: this should raise a django `CommandError`

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:150]: was this causing an error?

File: ppa/archive/tests/test_generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/tests/test_generate_textcorpus.py:216]: Don't forget to clean these up. Have these been replaced?

File: ppa/archive/tests/test_generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/tests/test_generate_textcorpus.py:378]: is this still a problem?

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:156]: to test this, just patch `progressbar` with a mock and you can confirm that is / is not called as expected based on the configuration

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (rlskoeser) [inline ppa/archive/management/commands/generate_textcorpus.py:2]: please add this to the sphinx docs - you can put it near where the other manage commands are listed

**Review by rlskoeser — COMMENTED** [review]
Revisions look good, looks like there is just some minor cleanup left.

Is there anything on our list that hasn't been addressed? I don't see any warning or summary counts, just the new error if the total from solr is zero. This is such a massive PR I'm inclined to get it merged sooner than later - any fine-tuning things that are left  could be done as a smaller PR once this is merged.

**Review by rlskoeser — APPROVED** [review]
Sorry, should have said - you have my ok to merge this once you deal with the remaining cleanup items in this PR, I don't need to review again

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:150]: yes! 😬 We needed to make sure the batch_size doesn't push us over the limit of what we want to pull from solr.

Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py:150]: is that a parasolr flaw?

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py:2]: should I remove `generate_corpus` or add `generate_textcorpus` beside it?

Comment (rlskoeser) [reply ppa/archive/management/commands/generate_textcorpus.py:2]: Is that Vineet's script? My inclination is to remove it. Can you imagine any scenario where we'd want to run that now that we have the new version?

File: ppa/archive/management/commands/generate_textcorpus.py
Comment (quadrismegistus) [reply ppa/archive/management/commands/generate_textcorpus.py]: not really, since solr doesn't know otherwise what our requested limit is. Just a mistake in how I thought resetting the batch_size might work, which is what I did earlier

**Pull request merged**
